File: app/src/main/java/com/github/mattiadellepiane/gnssraw/data/SharedData.java
package com.github.mattiadellepiane.gnssraw.data;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

import com.github.mattiadellepiane.gnssraw.MeasurementProvider;
import com.github.mattiadellepiane.gnssraw.R;
import com.github.mattiadellepiane.gnssraw.listeners.ServerCommunication;
import com.github.mattiadellepiane.gnssraw.ui.main.tabs.FilesFragment;

public class SharedData{
    //Singleton
    private SharedData(){}

    public static SharedData getInstance(){
        return BillPughSingleton.INSTANCE;
    }

    private static class BillPughSingleton{
        private static final SharedData INSTANCE = new SharedData();
    }

    //Inizio
    public MeasurementProvider measurementProvider;
    private boolean listeningForMeasurements = false;
    private ServerCommunication serverCommunication;
    private Context context;
    private FilesFragment filesFragment;

    public FilesFragment getFilesFragment(){
        return filesFragment;
    }

    public void setFilesFragment(FilesFragment filesFragment){
        this.filesFragment = filesFragment;
    }

    public Context getContext(){
        return context;
    }

    public void setContext(Context context){
        this.context = context;
    }

    public boolean isListeningForMeasurements() {
        return listeningForMeasurements;
    }

    public void setListeningForMeasurements(boolean listeningForMeasurements) {
        this.listeningForMeasurements = listeningForMeasurements;
    }

    public void setServerCommunication(ServerCommunication serverCommunication) {
        this.serverCommunication = serverCommunication;
    }
    public ServerCommunication getServerCommunication(){
        return serverCommunication;
    }

    public void startMeasurements(){
        setListeningForMeasurements(true);
        measurementProvider.registerAll();
    }

    public void stopMeasurements(){
        setListeningForMeasurements(false);
        measurementProvider.unRegisterAll();
    }

    //Preferences Getters
    public boolean isSensorsEnabled() {
        return PreferenceManager.getDefaultSharedPreferences(context).getBoolean("sensors_enabled", false);
    }
    public boolean isServerEnabled() {
        return PreferenceManager.getDefaultSharedPreferences(context).getBoolean("server_enabled", false);
    }
    public String getServerAddress() {
        return PreferenceManager.getDefaultSharedPreferences(context).getString("server_address", context.getString(R.string.server_ip_default));
    }
    public int getServerPort() {
        int port;
        try{
            port = Integer.parseInt(PreferenceManager.getDefaultSharedPreferences(context)
                    .getString("server_port", context.getString(R.string.server_port_default)));
        }catch(NumberFormatException | ClassCastException e){
            port = Integer.parseInt(context.getString(R.string.server_port_default));
        }
        return port;
    }
}

File: app/src/main/java/com/github/mattiadellepiane/gnssraw/data/SharedData.kt
package com.github.mattiadellepiane.gnssraw.data

import com.github.mattiadellepiane.gnssraw.R
import com.github.mattiadellepiane.gnssraw.ui.main.tabs.PlotFragment

import com.github.mattiadellepiane.gnssraw.ui.main.tabs.FilesFragment
import com.github.mattiadellepiane.gnssraw.ui.main.tabs.MapsFragment

import com.github.mattiadellepiane.gnssraw.MeasurementProvider
import com.github.mattiadellepiane.gnssraw.listeners.ServerCommunication
import java.lang.NumberFormatException
import java.lang.ClassCastException

import android.content.*
import android.util.Log
import androidx.preference.PreferenceManager

class SharedData{  //Singleton

    //Inizio
    var measurementProvider: MeasurementProvider? = null
    var isListeningForMeasurements = false
    var serverCommunication: ServerCommunication? = null
    var context: Context? = null
    var filesFragment: FilesFragment? = null
    var mapsFragment: MapsFragment? = null
    var plotFragment: PlotFragment? = null
    fun startMeasurements() {
        isListeningForMeasurements = true
        measurementProvider!!.registerAll()
    }

    fun stopMeasurements() {
        isListeningForMeasurements = false
        measurementProvider!!.unRegisterAll()
    }

    //Preferences Getters
    val isSensorsEnabled: Boolean
        get() = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("sensors_enabled", false)
    val isServerEnabled: Boolean
        get() = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("server_enabled", false)
    val serverAddress: String?
        get() = PreferenceManager.getDefaultSharedPreferences(context).getString("server_address", context!!.getString(R.string.server_ip_default))
    val fullTracking: Boolean
        get() = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("full_tracking", false)
    val serverPort: Int
        get() {
            val port: Int
            port = try {
                PreferenceManager.getDefaultSharedPreferences(context)
                        .getString("server_port", context!!.getString(R.string.server_port_default))!!.toInt()
            } catch (e: NumberFormatException) {
                context!!.getString(R.string.server_port_default).toInt()
            } catch (e: ClassCastException) {
                context!!.getString(R.string.server_port_default).toInt()
            }
            return port
        }

    companion object {
        val instance = SharedData();
    }
}